### PR TITLE
SAMZA-1835: Consolidate all processorId generation code.

### PR DESCRIPTION
**Pull request by shanthoosh**
Currently, the processorId creation function createProcessorId() is repeated in three different implementation of `JobCoordinator` viz `ZkJobCoordinator`, `PassthroughJobCoordinator`, and `AzureJobCoordinator`.  Here're the few problems that stems from this duplication.

1. `ProcessorId` is passed into the `MetricsReporterFactory` through the factory create method: `MetricsReporter getMetricsReporter(String name, String processorId, Config config);`. Custom `MetricsReporter` implementations currently use the processorId as a component in the generated metric names. Metrics reporters are instantiated from `LocalApplicationRunner` and`processorId` is currently passed in as null to `MetricsReporterFactory.getMetricsReporter`. This corrupts the generated metrics names.
2. `ZkJobCoordinator`, `ZkUtils`,  `ZkLeaderElector` and different downstream components of `LocalApplicationRunner` currently instantiate and manage their private reporters, rather than the sharing common `MetricsRegistry` managed by `LocalApplicationRunner`. Since there is no common namespace and reporter shared by downstream components of `LocalApplicationRunner`,  generating metrics dashboards for standalone is kind of a hassle.

This PR is comprised of the following changes:

1. Moved the processorId generation to `LocalApplicationRunner` and injects the generated `processorId` to all the downstream layers.
2. Deprecated the getProcessorId API in `JobCoordinator` interface.
3. Add the `processorId` and `metricsRegistry` arguments to the `getJobCoordinator` method of `JobCoordinatorFactory` interface.
4. Fixed the unit tests and added unit tests for `LocalApplicationRunner.createProcessorId`.

### CODE REVIEW

**Comment by shanthoosh** [comment]
@dnishimura @prateekm 
Please take a look when you get a chance.

**Comment by dnishimura** [comment]
Can you please address the build error reported from Travis-CI? Thanks.
```
/home/travis/build/apache/samza/samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java:208: error: no suitable constructor found for StreamProcessor(Config,String,Map<String,MetricsReporter>,TaskFactory,Optional<Object>,Optional<Object>,(sp)->proc[...]tener,JobCoordinator)
    this(config, processorId, customMetricsReporters, taskFactory, Optional.empty(), Optional.empty(), sp -> processorListener,
    ^
    constructor StreamProcessor.StreamProcessor(Config,String,Map<String,MetricsReporter>,TaskFactory,ProcessorLifecycleListener) is not applicable
      (actual and formal argument lists differ in length)
    constructor StreamProcessor.StreamProcessor(Config,String,Map<String,MetricsReporter>,TaskFactory,ProcessorLifecycleListener,JobCoordinator) is not applicable
      (actual and formal argument lists differ in length)
    constructor StreamProcessor.StreamProcessor(Config,String,Map<String,MetricsReporter>,TaskFactory,Optional<ApplicationContainerContextFactory<ApplicationContainerContext>>,Optional<ApplicationTaskContextFactory<ApplicationTaskContext>>,Optional<ExternalContext>,StreamProcessorLifecycleListenerFactory,JobCoordinator) is not applicable
      (actual and formal argument lists differ in length)```

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 1]
```diff
@@ -167,13 +169,36 @@ StreamProcessor createStreamProcessor(Config config, ApplicationDescriptorImpl<?
...
+   * 3. Else throws the {@see ConfigException} back to the caller.
+   * @param appConfig the configuration of the samza application.
+   * @throws ConfigException if neither processor.id nor app.processor-id-generator.class is defined in the configuration.
+   * @return the generated processor identifier.
+   */
+  @VisibleForTesting
+  static String createProcessorId(ApplicationConfig appConfig) {
+    if (appConfig.getProcessorId() != null) {
+      return appConfig.getProcessorId();
+    } else if (appConfig.getAppProcessorIdGeneratorClass() != null) {
```
Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: isNotBlank

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java [hunk 2]
```diff
@@ -167,13 +169,36 @@ StreamProcessor createStreamProcessor(Config config, ApplicationDescriptorImpl<?
       Optional<ExternalContext> externalContextOptional) {
     TaskFactory taskFactory = TaskFactoryUtil.getTaskFactory(appDesc);
     Map<String, MetricsReporter> reporters = new HashMap<>();
-    // TODO: the null processorId has to be fixed after SAMZA-1835
+    String processorId = createProcessorId(new ApplicationConfig(config));
```
[the remark this answers is not in the document]
Comment (shanthoosh) [reply samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: Sure.

**Comment by vjagadish1989** [comment]
I ended up merging this, @shanthoosh and @atoomula, please ensure you're in agreement on the changes.

File: samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (atoomula) [reply samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java]: This change looks fine. We will make the corresponding change in brooklin.

**Pull request closed without merging**